include/cudaq/Optimizer/Dialect/Common/Traits.h

```diff
@@ -13,10 +13,38 @@
 namespace cudaq {
 
 template <typename ConcreteType>
-class Hermitian : public mlir::OpTrait::TraitBase<ConcreteType, Hermitian> {};
+class Hermitian : public mlir::OpTrait::TraitBase<ConcreteType, Hermitian> {
+public:
+  static mlir::LogicalResult verifyTrait(mlir::Operation *op) {
+    if (cast<ConcreteType>(op).isAdj())
```

Comment:
I might be misinterpreting this check - I don't think we want to necessarily throw an error if the current op, though Hermitian, is adjoint. It is ok for the Hermitian op to be marked adjoint, it's just that it's equivalent to not being marked adjoint.

Reply:
What would the purpose be of allowing these ops to be flagged with the adjoint flag?

Allowing ops in the IR that are already self-adjoint the extra expressiveness of adding a flag with no semantics violates the principle of minimalism. Disallowing it in the IR simplifies all clients as they don't have to check for themselves that the flag is noise and has no semantics.

This change is to the Quake IR, so it doesn't impact the C++ syntax, where the rules could be more fast and loose.

Reply:
There is no purpose for the flag on self-adjoint ops, and your reasoning makes sense for disallowing it at the IR level. So I'm cool with this.